site/en/blog/crx-scripting-api/index.md

```diff
@@ -0,0 +1,306 @@
+frame in the tab.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.webNavigation.getAllFrames({tabId: tab.id}, (frames) => {
+    let frame1 = frames[0].frameId;
+    let frame2 = frames[1].frameId;
+
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame1,
+      file: 'content-script.js',
+    });
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame2,
+      file: 'content-script.js',
+    });
+  });
+});
+```
+
+In Manifest V3, we replaced the optional `frameId` integer property in the options object with an
+optional `frameIds` array of integers; this allows developers to target multiple frames in a single
+API call. You can use use `chrome.webNavigation.getAllFrames` to get a list of all frames in a tab.
+
+```js
+// Manifest V3 extension
+chrome.action.onClicked.addListener(async (tab) => {
+  let frames = await new Promise((resolve) => {
+    chrome.webNavigation.getAllFrames({tabId: tab.id}, resolve);
+  });
+  let frame1 = frame[0].frameId;
+  let frame2 = frame[1].frameId;
+
+  chrome.scripting.executeScript({
+    target: {
+      tabId: tab.id,
+      frameIds: [frame1, frame2],
+    },
+    files: ['content-script.js'],
+  });
+});
+```
+
+### Script injection results
+
+We've also improved the way we return script injection results in Manifest V3. A "result" is
+basically the final statement evaluated in a script. Think of it like the value returned when you
+call `eval()` or execute a block of code in the Chrome DevTools console, but serialized in order to
+pass results across processes.
+
+In Manifest V2, `executeScript` and `insertCSS` would return an array of plain execution results.
+This is fine if you only have a single injection point, but result order is not guaranteed when
+injecting into multiple frames so there's no way to tell which result is associated with which
+frame.
+
+For a concrete example, let's take a look at the `results` arrays returned by a Manifest V2 and a
+Manifest V3 version of the same extension. Both versions of the extension will inject the same
+content script and we'll be comparing results on the same [demo page][iframe-demo].
+
+```js
+//// content-script.js
+var headers = document.querySelectorAll('h1,h2,h3,h4,h5,h6');
+headers.length;
+```
+
+When we run the Manifest V2 version, we get back an array of `[0, 1]`. Which result
+corresponds to the main frame and which is for the iframe? The return value doesn't tell us, so we
+don't know for sure.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.tabs.executeScript({
+    allFrames: true,
+    file: 'content-script.js',
+  }, (results) => {
+    // results == [1, 0]
+    for (let result of results) {
+      if (result > 0) {
+        // Do something with the frame… which one was it?
+      }
+    }
+  });
+});
+```
+
+In the Manifest V3 version, results now contains an array of objects instead of an array of strings,
```

Comment:
```suggestion
In the Manifest V3 version, `results` now contains an array of result objects instead of an array of just the evaluation results,
```

(results could be non-strings)

Reply:
Thanks for the clarification.